rfcs/react-components/styles-handbook.md

```diff
@@ -0,0 +1,771 @@
+
+- selector `.fqhvij7 > .ui-button`
+- match_attempts **1**
+- match_count **1**
+
+# Best practices
+
+## Writting styles
+
+### Use `tokens` over direct colors
+
+Fluent UI React v9 provides [design tokens][fluent-colors] to provide consistent way of theming via `tokens` provided by `@fluentui/react-theme`.
+
+```js
+import { makeStyles } from '@griffel/react';
+import { tokens } from '@fluentui/react-theme';
+
+const useClasses = makeStyles({
+  // ❌ Don't do
+  rootA: { color: 'red' /* brand foreground */ },
+  // ✅ Do
+  rootB: { color: tokens.colorBrandForeground1 },
+});
+```
+
+Exact colors should be never used in Fluent UI code and are not recommended by be used in applications. The exception are [system-colors][mdn-system-colors] used by forced colors mode:
+
+```js
+import { makeStyles } from '@griffel/react';
+
+const useClasses = makeStyles({
+  button: { color: 'ButtonText' },
+});
+```
+
+### Avoid duplication in rule definitions
+
+Styles written for components should follow these rules:
+
+- base styles should contain the most of CSS definitions that are applicable to base state
+- styles for permutations should be granular and don't contain definitions that are present in base styles
+
+```js
+import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
+import { tokens } from '@fluentui/react-theme';
+
+const useClasses = makeStyles({
+  base: {
+    display: 'flex',
+    color: tokens.colorNeutralForeground1,
+    ...shorthands.padding('10px'),
+  },
+  // ❌ Don't do
+  //  "display" & "padding" with the same values are defined in base styles
+  primary: {
+    display: 'flex',
+    ...shorthands.padding('10px'),
+    backgroundColor: tokens.colorBrandBackground,
+    color: tokens.colorBrandForeground1,
+  },
+  // ✅ Do
+  primary: {
+    backgroundColor: tokens.colorBrandBackground,
+    color: tokens.colorBrandForeground1,
+  },
+});
+
+function App(props) {
+  const classes = useClasses();
+  const className = mergeClasses(classes.base, props.primary && classes.primary);
+
+  /* --- */
+}
+```
+
+### Avoid `!important`
+
+Our styles are written in way to allow predictable and simple style overrides, `!important` should not be used in styles:
+
+```js
+import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
+
+const useClasses = makeStyles({
+  base: {
+    // ❌ Don't do
+    display: 'flex !important',
+  },
+});
+```
+
+## Performance
+
+### Avoid nested `mergeClasses` calls
+
+`mergeClasses` is a performant function, however it's not expected that it will be called multiple times for the same element.
+
+```js
+// ❌ Don't do
+function Component(props) {
+  const classes = useClasses();
+
+  const classesForFoo = mergeClasses(/* ---- */);
+  const className = mergeClasses(classes.root, classesForFoo, mergeClasses(/* ---- */), mergeClasses(/* ---- */));
+
+  /* --- */
+}
+```
+
+Conditions to apply styles might be complex, in this case consider to extract them to separate variables:
```

Comment:
exctracting coniditions to separate variables should be it's own section IMO, maybe also with tricks like

```ts
makeStyesl({
  small: {},
  medium: {},
  large: {},
})

mergeClasses(styes[props.size]);
```

something like 'structure your conditional styles'

Reply:
Good idea, added in ddbdb92.